pkg/controller/kfservice/resources/istio/virtualservice.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2019 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package istio
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha2"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	istiov1alpha1 "knative.dev/pkg/apis/istio/common/v1alpha1"
+	istiov1alpha3 "knative.dev/pkg/apis/istio/v1alpha3"
+)
+
+const (
+	IngressConfigKeyName = "ingress"
+)
+
+// Status Constants
+var (
+	PredictorSpecMissing       = "predictorSpecMissing"
+	PredictorStatusUnknown     = "predictorStatusUnknown"
+	PredictorHostnameUnknown   = "predictorHostnameUnknown"
+	TransformerStatusUnknown   = "transformerStatusUnknown"
+	TransformerHostnameUnknown = "transformerHostnameUnknown"
+	ExplainerSpecMissing       = "explainerSpecMissing"
+	ExplainerStatusUnknown     = "explainerStatusUnknown"
+	ExplainerHostnameUnknown   = "explainerHostnameUnknown"
+)
+
+type IngressConfig struct {
+	IngressGateway     string `json:"ingressGateway,omitempty"`
+	IngressServiceName string `json:"ingressService,omitempty"`
+}
+
+type VirtualServiceBuilder struct {
+	ingressConfig *IngressConfig
+}
+
+func NewVirtualServiceBuilder(config *corev1.ConfigMap) *VirtualServiceBuilder {
+	ingressConfig := &IngressConfig{}
+	if ingress, ok := config.Data[IngressConfigKeyName]; ok {
+		err := json.Unmarshal([]byte(ingress), &ingressConfig)
+		if err != nil {
+			panic(fmt.Errorf("Unable to unmarshall ingress json string due to %v ", err))
+		}
+
+		if ingressConfig.IngressGateway == "" || ingressConfig.IngressServiceName == "" {
+			panic(fmt.Errorf("Invalid ingress config, ingressGateway and ingressService are required."))
+		}
+	}
+
+	return &VirtualServiceBuilder{ingressConfig: ingressConfig}
+}
+
+func createFailedStatus(reason string, message string) *v1alpha2.VirtualServiceStatus {
+	return &v1alpha2.VirtualServiceStatus{
+		Status: duckv1beta1.Status{
+			Conditions: duckv1beta1.Conditions{{
+				Type:    v1alpha2.RoutesReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  reason,
+				Message: message,
+			}},
+		},
+	}
+}
+
+func (r *VirtualServiceBuilder) CreateVirtualService(kfsvc *v1alpha2.KFService) (*istiov1alpha3.VirtualService, *v1alpha2.VirtualServiceStatus) {
```

Comment:
I find this interface to be somewhat awkward. Without digging too deeply, I'd expect the virtual service to be statically 
 configurable and applied in the reconciler, so it seems odd to me to return a status here. 

I'd need to take a deeper look as to why this structure might be necessary. Let's do this later, but I was hoping for this class to be overall simpler w/ less logic.

Reply:
This is because the creation and shape is dependent on so many things... an alternate would be to identify all the cases in the reconciler and then ask this function to build exactly what is asked... but would rather not make that change at this time...